x-pack/plugins/files/server/blob_storage_service/adapters/es/es.ts

```diff
@@ -36,6 +38,9 @@ export class ElasticsearchBlobStorage implements BlobStorage {
     );
   }
 
+  // TODO: We should use a MutEx here to ensure that the initial creation request of this
+  // index only happens once. Otherwise we can hit an edge case where if two files are being
+  // uploaded in the same instant one of the requests to upload can fail.
```

Comment:
Maybe not mutex, but we can just store the creation promise, e.g.:

```ts
__createIndexIfNotExistsPromise = null;
createIndexIfNotExists() {
  if (!this.__createIndexIfNotExistsPromise) {
    this.__createIndexIfNotExistsPromise = // ...
  }
  return this.__createIndexIfNotExistsPromise;
}
```

---

Also, that would just help locally. Two Kibana server nodes could still try to store a file at the same time.

Reply:
Yeah, multiple Kibana's is a tricky one. The lock would need to live in ES where it is visible to all. I guess it is not worth the added overhead. I'll remove this comment.